src/mutable_data.rs

```diff
@@ -68,48 +127,537 @@ impl MutableData {
+        self.permissions.get(user).ok_or(DataError::NoSuchKey)
+    }
 
-#[derive(Hash, Eq, PartialEq, PartialOrd, Ord, Clone, Serialize, Deserialize)]
-pub struct Permissions {
-    permissions: BTreeMap<User, BTreeSet<Permission>>,
-    /// The current index of the data when this permission change happened.
-    data_index: u64,
-    /// The current index of the owners when this permission change happened.
-    owner_entry_index: u64,
+    /// Insert or update permissions for the provided user.
+    pub fn set_user_permissions(
+        &mut self,
+        user: PublicKey,
+        permissions: PermissionSet,
+        version: u64,
+        requester: PublicKey,
+    ) -> Result<(), DataError> {
+        if !self.is_action_allowed(requester, Action::ManagePermissions) {
+            return Err(DataError::AccessDenied);
+        }
+        if version != self.version + 1 {
+            return Err(DataError::InvalidSuccessor(self.version));
+        }
+        let _prev = self.permissions.insert(user, permissions);
+        self.version = version;
+        Ok(())
+    }
+
+    /// Delete permissions for the provided user.
+    pub fn del_user_permissions(
+        &mut self,
+        user: &PublicKey,
+        version: u64,
+        requester: PublicKey,
+    ) -> Result<(), DataError> {
+        if !self.is_action_allowed(requester, Action::ManagePermissions) {
+            return Err(DataError::AccessDenied);
+        }
+        if version != self.version + 1 {
+            return Err(DataError::InvalidSuccessor(self.version));
+        }
+        if !self.permissions.contains_key(user) {
+            return Err(DataError::NoSuchKey);
+        }
+        let _ = self.permissions.remove(user);
+        self.version = version;
+        Ok(())
+    }
+
+    /// Delete user permissions without performing any validation.
+    pub fn del_user_permissions_without_validation(
+        &mut self,
+        user: &PublicKey,
+        version: u64,
+    ) -> bool {
+        if version <= self.version {
+            return false;
+        }
+        let _ = self.permissions.remove(user);
+        self.version = version;
+        true
+    }
+
+    /// Change owner of the mutable data.
+    pub fn change_owner(&mut self, new_owner: PublicKey, version: u64) -> Result<(), DataError> {
+        if version != self.version + 1 {
+            return Err(DataError::InvalidSuccessor(self.version));
+        }
+        self.owners = new_owner;
+        self.version = version;
+        Ok(())
+    }
+
+    /// Change the owner without performing any validation.
+    pub fn change_owner_without_validation(&mut self, new_owner: PublicKey, version: u64) -> bool {
+        if version <= self.version {
+            return false;
+        }
+
+        self.owners = new_owner;
+        self.version = version;
+        true
+    }
+
+    pub fn is_action_allowed(&self, requester: PublicKey, action: Action) -> bool {
+        if self.owners == requester {
```

Comment:
This wouldn't be correct really with BLS, since `PublicKey` can be comprised of multiple keys. So it should take a signature for an owner instead. But we'll deal with it as a separate task, so OK to leave it as such for now.

Reply:
Yea. That was my understanding too. This is meant to be just a temporary placeholder until we properly implement BLS signature verification